stdlib/public/core/Random.swift

```diff
@@ -0,0 +1,167 @@
+
+    if remainder != 0 {
+      let random = next()
+      let mask = UInt64.max &>> (UInt64.bitWidth - remainder)
+      tmp += T(truncatingIfNeeded: random & mask) &<< (UInt64.bitWidth * quotient)
+    }
+
+    return tmp
+  }
+
+  /// Returns a random value that is less than the given upper bound.
+  ///
+  /// - Parameter upperBound: The upper bound for the randomly generated value.
+  /// - Returns: A random value of `T` in the range `0..<upperBound`. Every
+  ///   value in the range `0..<upperBound` is equally likely to be returned.
+  @inlinable
+  public mutating func next<T: FixedWidthInteger & UnsignedInteger>(
+    upperBound: T
+  ) -> T {
+    guard upperBound != 0 else { return 0 }
+    let tmp = (T.max % upperBound) + 1
+    let range = tmp == upperBound ? 0 : tmp
+    var random: T = 0
+
+    repeat {
+      random = next()
+    } while random < range
+
+    return random % upperBound
+  }
+}
+
+/// The default source of random data.
+///
+/// When you generate random values, shuffle a collection, or perform another
+/// operation that depends on random data, this type's `default` property is
+/// the generator used by default. For example, the two method calls in this
+/// example are equivalent:
+///
+///     let x = Int.random(in: 1...100)
+///     let y = Int.random(in: 1...100, using: &Random.default)
+///
+/// `Random.default` is safe to use in multiple threads, and uses a
+/// cryptographically secure algorithm whenever possible.
+///
+/// Platform Implementation of `Random`
+/// ===================================
+///
+/// - Apple platforms use `arc4random_buf(3)` for newer versions of their OS,
+///   and for older versions they will read from `/dev/urandom`.
+/// - `Linux`, `Android`, `Cygwin`, `Haiku`, `FreeBSD`, and `PS4` all try to
+///   use `getrandom(2)`, but if it doesn't exist then they read from
+///   `/dev/urandom`.
+/// - `Fuchsia` calls `getentropy(3)`.
+/// - `Windows` calls `BCryptGenRandom`.
+public struct Random : RandomNumberGenerator {
+  /// The default instance of the `Random` random number generator.
+  public static var `default`: Random {
+    get { return Random() }
+    set { /* Discard */ }
+  }
+
+  private init() {}
+
+  /// Returns a value from a uniform, independent distribution of binary data.
+  ///
+  /// - Returns: An unsigned 64-bit random value.
+  public mutating func next() -> UInt64 {
+    var random: UInt64 = 0
+    _stdlib_random(&random, MemoryLayout<UInt64>.size)
+    return random
+  }
+  
+  /// Returns a value from a uniform, independent distribution of binary data.
+  ///
+  /// - Returns: A random value of `T`. Bits are randomly distributed so that
+  ///   every value of `T` is equally likely to be returned.
+  public mutating func next<T: FixedWidthInteger & UnsignedInteger>() -> T {
+    var random: T = 0
+    _stdlib_random(&random, MemoryLayout<T>.size)
+    return random
+  }
+}
+
+public // @testable
+func _stdlib_random(_ bytes: UnsafeMutableRawBufferPointer) {
```

Comment:
If filling a buffer with random bytes is the primitive operation around which we're building the default RNG, it should probably be exposed as a customizable public operation on `RandomNumberGenerator`. (I.e., I think it should be a requirement, with a default implementation in terms of the 64-bit `next`.)

(Update: I see you've already made the case for this on swift-evolution. 👍)

Reply:
Yeah I see that as a good future enhancement.